ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/internal/EditorConfigFinder.kt

```diff
@@ -0,0 +1,85 @@
+package com.pinterest.ktlint.core.internal
+
+import com.pinterest.ktlint.core.initKtLintKLogger
+import com.pinterest.ktlint.core.internal.ThreadSafeEditorConfigCache.Companion.threadSafeEditorConfigCache
+import java.nio.charset.StandardCharsets
+import java.nio.file.FileVisitResult
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.SimpleFileVisitor
+import java.nio.file.attribute.BasicFileAttributes
+import kotlin.io.path.isDirectory
+import mu.KotlinLogging
+import org.ec4j.core.Resource
+import org.ec4j.core.ResourcePropertiesService
+import org.ec4j.core.model.Version
+import org.jetbrains.kotlin.konan.file.File
+
+private val logger = KotlinLogging.logger {}.initKtLintKLogger()
+
+internal class EditorConfigFinder {
+    /**
+     * Finds all relevant ".editorconfig" files for the given path.
+     */
+    fun findEditorConfigs(path: Path): List<Path> {
+        val result = mutableListOf<Path>()
+        val normalizedPath = path.normalize().toAbsolutePath()
+        if (path.isDirectory()) {
+            result += findEditorConfigsInSubDirectories(normalizedPath)
+        }
```

Comment:
I started wondering if there shouldn't be a way to exclude searching in subdirectories 🤔 I initially thought of passing project root directory as the `path`, but that would result in traversing through all files and folders, including `build`, `docs` or `resources` directory, which would be wasteful.

Gradle plugins would have to be smarter, and we'd have to figure out a way to identify the root folder for all directories containing Kotlin sources, which unfortunately I failed to do so :/ For Kotlin plugin we get [SourceDirectorySet#getSrcDirTrees](https://docs.gradle.org/current/javadoc/org/gradle/api/file/SourceDirectorySet.html#getSrcDirTrees--) which would allow us to identify multiple root directories, but I'm unaware of a similar way to obtain such information for other plugins (for example, for Android Gradle Plugin).

Given all of the above, I started leaning towards ignoring subdirectories, for performance reasons (and I believe sharing common `editorconfig` settings for a compilation unit is a fairly common situation). Having the subdirectories' traversal optional, `ktlint` consumers could switch to a more precise input calculation if they are able to identify a _proper top-most path_. 
Do you have any thoughts/ideas/advices here? 

I can link my comment on the same topic: https://github.com/jeremymailen/kotlinter-gradle/pull/265#discussion_r965632248

Reply:
Why do you think the performance impact of scanning all subdirectories it too big? This call should be made only once. For testing KtLint I always run KtLint on a collection of open source projects which in total contain 5,000+ directories and more than 23,000 files and the scanning is really fast.

I have not encountered many use cases in which a project contains multiple `.editorconfig` files. But I think it would be really confusing for developers if the `.editorconfig` file in the root of the project and its parents is treated differently from `.editorconfig` files in subdirectories.

As with all performance issues, my advice would be first to measure before deciding to optimize.

Reply:
>Why do you think the performance impact of scanning all subdirectories it too big?

I don't know. I haven't ever done any sort of file performance things. My assumption was that _doing_ things takes more time that completely avoiding it 😛 

>This call should be made only once

From Gradle Plugin perspective, it's once per task invocation, so every time someone calls `./gradlew ktlintTask`. 

>As with all performance issues, my advice would be first to measure before deciding to optimize.

Sure, sounds reasonable 👍  I'll make sure to do extra profiling when switching to the new API 👍 Thanks for answering my doubts (and for finding time to work on the new api) 🙏